Speedup headers parsing

m-burst opened the pull request:
Replace `_CRLF_RE.split(headers)` with a simple `headers.split('\n')` with an additional check for '\r' in each line.

I've benchmarked this change on Tornado 4.5 with the latest pypy2.7 (that's what we have in production now): the splitting itself is ~20x faster, and the whole headers parsing is ~4-5x faster.

I'm opening this as a draft PR to discuss the feasibility of this change; if you like the general idea, I'm willing to add any necessary changes.

bdarnell commented:
That's an impressive (and surprising) speedup! I'd be happy to see a change like this as long as the speedup holds across python versions. We need to benchmark this on Python 3 (since Tornado no longer supports Python 2), and also compare both cpython and pypy. Please add a script in the `maint/benchmark` directory that we can use to measure this. 

I kind of expect that the regex version will be faster in cpython - the extra `line.endswith` call on every loop will negate any benefit from using `str.split` instead of `re.split`. In that case we'll have to figure out what to do - is it worth taking a performance hit on cpython, or maintaining two slightly different versions? It depends in part on how the numbers turn out.

It's also possible that there's an inefficiency that could be fixed in pypy's `re` module. I wouldn't expect `re.split` to be 20x slower than `str.split` for this case (although maybe i'm missing some optimizations that are possible for `str.split`).

m-burst commented:
Added a simple benchmark.

Here are the results for `headers-split`:

```
Python 3.5.7
regex 1.2804281000007904
str.split 0.4018657999995412
speedup 3.186208181940967

Python 3.6.8
regex 1.3048213000001851
str.split 0.4044023999995261
speedup 3.2265419295279014

Python 3.7.3
regex 1.374576899999738
str.split 0.4354571000003489
speedup 3.1566298953413248

Python 3.5.3 (3f6eaa010fce78cc7973bdc1dfdb95970f08fed2, Jan 14 2018, 12:54:44)
[PyPy 5.10.1 with GCC 7.2.0]
regex 1.6284960000002684
str.split 0.11339799999950628
speedup 14.360888199151296

Python 3.6.1 (784b254d669919c872a505b807db8462b6140973, Apr 16 2019, 18:18:28)
[PyPy 7.1.1-beta0 with GCC 8.2.0]
regex 1.761060799999541
str.split 0.09382079999977577
speedup 18.770473072109272
```

And for `headers-full`:

```
Python 3.5.7
regex 0.43024960000002466
str.split 0.34017560000029334
speedup 1.2647867748293928

Python 3.6.8
regex 0.4392166999996334
str.split 0.35676279999916005
speedup 1.2311168653252735

Python 3.7.3
regex 0.36412860000018554
str.split 0.27547580000009475
speedup 1.3218170162317717

Python 3.5.3 (3f6eaa010fce78cc7973bdc1dfdb95970f08fed2, Jan 14 2018, 12:54:44)
[PyPy 5.10.1 with GCC 7.2.0]
regex 0.2187298999997438
str.split 0.06846700000005512
speedup 3.194676267392579

Python 3.6.1 (784b254d669919c872a505b807db8462b6140973, Apr 16 2019, 18:18:28)
[PyPy 7.1.1-beta0 with GCC 8.2.0]
regex 0.2401668000002246
str.split 0.05526190000000497
speedup 4.3459743512293825
```

CPython tests were ran with `--num-runs=1`, PyPy tests were ran with `--num-runs=15` and the last run result is included.

m-burst commented:
@bdarnell Please take a look at the benchmark and its results in the comment above. Thanks!

m-burst commented:
@bdarnell Thanks for the review! I've squashed the commits, rebased onto master and removed the draft flag.
